rust/kernel/chrdev.rs

```diff
@@ -20,10 +19,57 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+/// Character device.
+///
+/// # Invariants
+///
+/// The pointer [`Cdev::cdev`] is valid and non-null.
+struct Cdev(*mut bindings::cdev);
+
+impl Cdev {
+    fn alloc(
+        fops: &'static bindings::file_operations,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult<Self> {
+        // SAFETY: `bindings::cdev_alloc` returns a valid pointer or null.
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        // SAFETY: [`Cdev::cdev`] is valid and non-null.
+        // `fops` and `module` have `'static` lifetime.
```

Comment:
Is that true, though? If `fops` or `module` get deallocated before the `struct cdev` they're part of... very interesting and unsound things could happen.

Reply:
Yeah, that is what I meant by "for future usage of this struct such as other FFI calls etc.". But for this `unsafe` block in particular, it is not making it sound or otherwise, i.e. even if both were null the block would still be safe.

Reply:
True! I'm not sure where the crashes could happen in such a case. If `owner` is freed prematurely, the call to `cdev_del` in `Drop` could crash. But worse, if `fops` is freed prematurely, the crash could happen anywhere in the kernel's C code!

I'm not sure how this could/should be documented or analysed.

Reply:
My suggestion (sorry if it was not clear) is to tell this fact in the type `# Invariants` section, since it is something that the type guarantees. Then perhaps you can mention the fields are properly filled and will never change thanks to the invariant when calling `cdev_add()`.